### ISSUE TITLE
advanced scene tree search

### ISSUE THREAD

**Issue by GuyUnger** [godotengine/godot-proposals#4986]
### Describe the project you are working on

NA

### Describe the problem or limitation you are having in your project

I have a lot of nodes in my scene and searching through them is limited

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

add advanced search filters similar to discord's search

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

`filter:script`
all nodes that have a script

`filter:tool`
all nodes with a tool script

`script:SomeScript`
all nodes that have SomeScript.gd attached

`filter:signal`
all nodes that have connected signals

`filter:group`
all nodes that are in a group

`group:group_name`
all nodes in a specific group

`type:AudioStreamPlayer2D`
all the AudioStreamPlayer2D's (or whatever node type)

i don't know the best syntax, but some way to filter out these kinds of properties would be very helpful for me

### If this enhancement will not be used often, can it be worked around with a few lines of script?

NA

### Is there a reason why this should be core and not an add-on in the asset library?

NA

**Comment by Zireael07** [comment]
@and-rad: I imagine the idea is to filter through nodes in a large scene while you edit it, not to find stuff at runtime. Your proposals are all for runtime.

**Comment by and-rad** [comment]
As in, typing it in the `Filter Nodes` search bar in the editor? That would indeed be a different use case and I misunderstood. (All the more reason to be more explicit in the proposal, though).

In that case, I'd be interested in how instanced scenes should be handled. Their children are hidden if `Editable Children` is not checked, so would they just be ignored by the search? It's what I would expect, but I imagine that would limit the usefulness of this feature somewhat.

**Comment by KoBeWi** [comment]
`type:` is being implemented here: https://github.com/godotengine/godot/pull/58377

**Comment by and-rad** [comment]
Makes sense. I think this proposal is a good idea then. I can see `group:` being pretty useful as well as the `filter:` filter. The question would then be what kinds of filters should be supported in the initial implementation.

**Comment by seppoday** [comment]
> It would probably be a good idea if you went into more detail about specific use cases where the current implementation is a limiting factor for you. Like, limiting in terms of the amount of code you have to write or in terms of performance or something else?


As @KoBeWi said, we already have search filter for "type: ". This PR is nice QoL change that does not remove or add anything visually to the editor. If someone will not use it that proposal would not change anything for that person. Editor will be the same. 

At the same time if someone needs that filters then why not have them? I think that proposal is great and we already have tested it with that "type:" PR. I think that it is especially nice feature for more complex scenes (UI heavy etc.) If there is contributor interested in implementing then let's do it :)

**Comment by GuyUnger** [comment]
> As in, typing it in the `Filter Nodes` search bar in the editor? That would indeed be a different use case and I misunderstood. (All the more reason to be more explicit in the proposal, though).
> 
> In that case, I'd be interested in how instanced scenes should be handled. Their children are hidden if `Editable Children` is not checked, so would they just be ignored by the search? It's what I would expect, but I imagine that would limit the usefulness of this feature somewhat.

right, sorry for being unclear. Currently instanced children do not show up in search, maybe it could be helpful but so far i haven't felt a need for that in the search functions currently

**Comment by and-rad** [comment]
If more filters are going be added besides the `type:` filter that's already being worked on, it might be a good idea to implement a kind of drop-down element for the search field for users to select among the available filters. Like how Discord does it.

**Comment by GuyUnger** [comment]
> If more filters are going be added besides the `type:` filter that's already being worked on, it might be a good idea to eventually implement a kind of drop-down element for the search field for users to select among the available filters. Like how Discord does it.

yes definitely agree

**Comment by GuyUnger** [comment]
instead of `filter:` it could also be `has:` like on discord

**Comment by JoNax97** [comment]
To streamline the search, we can use wildcards instead of the 'filter' keyword.
For example, instead of this:

> filter:group
all nodes that are in a group
> group:group_name
all nodes in a specific group

We can have this:

> group: *
all nodes that are in a group
group: group_name
all nodes in a specific group

This way the syntax is the same for both cases

**Comment by me2beats** [comment]
Also there can be a way to search nodes in the project, for all scenes.
Maybe worth a separate proposal tho

**Comment by GuyUnger** [comment]
this can also greatly improve some of the prompts for selecting nodes such as:
connecting signal: default search to `filter:script`
adding audio playback track to animator `type:AudioStreamPlayer2D,AudioStreamPlayer3D,AudioStreamPlayer` (don't know the appropriate syntax for allowing multiple types)
adding a call method track to animator `filter:script`

**Comment by seppoday** [comment]
Other idea for possible implementation (if there is need for it)

- filter:uniquename

**Comment by Mickeon** [comment]
If I may suggest, `filter:unique` or something similar, to only highlight **Scene Unique Nodes**. Somewhat needed, as it can prove difficult to spot and make use of them by memory in a large Scene Tree.

**Comment by d4rkd0s** [comment]
Just FYI there's also #4040 which was worked on, and implemented here: https://github.com/godotengine/godot/pull/58377
Just posting to link/relate these two "issues" together <3
